### PR TITLE
Implement `smart_sampling` for `dq.jssesolve()`

### PR DESCRIPTION
**Pull request by dkweiss31**
Follow up on https://github.com/dynamiqs/dynamiqs/pull/748, reimplementing the smart sampling algorithm whereby the no-jump trajectory is only sampled once. I want to offer my thanks again to @gautierronan for your heroic simplification of my initial overcomplicated implementation, which I'm only now internalizing upon diving into this reimplementation. A few notes/questions

1. In this implementation, the trajectories are the final (rightmost) batch dimension (this is actually consistent with the previous version of the API docstring) as opposed to previously where they were the leftmost. I like them being leftmost, because it is natural to sum/average over `axis=0`. However this would require some inelegant `moveaxis` calls at the end of `_vectorized_jssesolve`. Thoughts?
2. To keep things simple, the no-jump and jump results are returned together, with e.g. `result.states[..., 0, :, :]` containing the no-click trajectory and `result.states[..., 1:, :, :]` containing the click trajectories (similarly for expectation values). Actually calculating averaged quantities is thus a little complicated for the user. They must call e.g. (if there are no other batch dimensions)
```python
noclick_prob = jsseresult.final_state_norm[0]
mean_jsse_expects_click = jnp.mean(jsseresult.expects[1:], axis=0)
mean_jsse_expects = (
      noclick_prob_expect * jsseresult.expects[0]
      + (1 - noclick_prob_expect) * mean_jsse_expects_click
)
```
to calculate expectation values. I wonder if this is something we can implement in the result class? We'd need to worry then also about batch dimensions.

### CODE REVIEW

**Review by gautierronan — CHANGES_REQUESTED** [review]
Great PR @dkweiss31! I love how minimal the actual logic is.

I have a few minor comments below. Otherwise, regarding your remarks:

> In this implementation, the trajectories are the final (rightmost) batch dimension (this is actually consistent with the previous version of the API docstring) as opposed to previously where they were the leftmost. I like them being leftmost, because it is natural to sum/average over axis=0. However this would require some inelegant moveaxis calls at the end of _vectorized_jssesolve. Thoughts?

For now, the trajectories dimension is always rightmost in other stochastic solvers. We should align this PR accordingly. However, we can definitely think about changing to leftmost globally in a separate PR.

That being said, I think it makes more sense to have it rightmost for now, because we treat `ntrajs` similarly to `ntsave`: it cannot be flat-batched over and only accepts a single batching dimension. I think this is meant to change eventually, and we should accept multiple batching dimensions + flat-batching.

> To keep things simple, the no-jump and jump results are returned together, with e.g. result.states[..., 0, :, :] containing the no-click trajectory and result.states[..., 1:, :, :] containing the click trajectories (similarly for expectation values). Actually calculating averaged quantities is thus a little complicated for the user. They must call e.g. (if there are no other batch dimensions) to calculate expectation values. I wonder if this is something we can implement in the result class? We'd need to worry then also about batch dimensions.

I rather like this API. Indeed we can definitely add a helper in the results class to compute mean values. I think it could take the form of a function with an API like:
```python
mean_states = jsseresult.mean(jsseresult.states, axis=-3)
mean_expects = jsseresult.mean(jsseresult.expects, axis=-1)
```
maybe with additional properties to make it even simpler
```python
mean_states = jsseresult.mean_states
mean_expects = jsseresult.mean_expects
```

File: dynamiqs/integrators/core/save_mixin.py [hunk 1]
```diff
@@ -83,10 +83,13 @@ def save(self, y: PyTree) -> Saved:
...
-        ylast = saved.ysave[-1]
-        saved = super().postprocess_saved(saved, ylast)
-        return JumpSolveSaved(saved.ysave, saved.extra, saved.Esave, clicktimes)
+    def postprocess_saved(
+        self, saved: SolveSaved, ylast: PyTree, clicktimes: Array
+    ) -> JumpSolveSaved:
+        saved = super().postprocess_saved(saved, unit(ylast))
+        return JumpSolveSaved(
+            saved.ysave, saved.extra, saved.Esave, clicktimes, ylast.norm()
+        )
```
[the remark this answers is not in the document]
Comment (dkweiss31) [reply dynamiqs/integrators/core/save_mixin.py:92]: This is necessary because the `save` method above forces normalization of the saved state, so we lose all information about the norm. So I need to be able to access the unnormalized `ylast` to extract the noclick_probability.

File: tests/jssesolve/test_jssesolve.py [hunk 2]
```diff
@@ -8,7 +8,8 @@
 
 
 @pytest.mark.run(order=TEST_LONG)
-def test_against_mesolve_oscillator(atol=5e-2):
+@pytest.mark.parametrize('smart_sampling', [True, False])
+def test_against_mesolve_oscillator(smart_sampling, atol=1e-1):
```
[the remark this answers is not in the document]
Comment (dkweiss31) [reply tests/jssesolve/test_jssesolve.py]: You are right! I had a bad bug, the noclick_probability is norm**2, not norm! Thanks for drawing attention to this :)

File: dynamiqs/integrators/apis/jssesolve.py [hunk 3]
```diff
@@ -295,12 +299,59 @@ def _vectorized_jssesolve(
...
+        # consume the remaining keys for the click trajectories, using the norm of the
+        # no-click state as the minimum value for random numbers triggering a click
+        click_args = (keys[1:], False, noclick_result.final_state_norm)
+        click_result = f(
+            H, Ls, psi0, tsave, *click_args, exp_ops, method, gradient, options
+        )
+
+        def _concatenate_results(x: PyTree, y: PyTree) -> PyTree:
+            if isinstance(x, float | int | bool) or len(x.shape) == len(y.shape):
+                return x
```
[the remark this answers is not in the document]
Comment (dkweiss31) [reply dynamiqs/integrators/apis/jssesolve.py:337]: This worked, with the slight modification to `return jnp.concatenate((leaf1[None], leaf2))`

**Comment by dkweiss31** [comment]
> 
> I rather like this API. Indeed we can definitely add a helper in the results class to compute mean values. I think it could take the form of a function with an API like:
> 
> ```python
> mean_states = jsseresult.mean(jsseresult.states, axis=-3)
> mean_expects = jsseresult.mean(jsseresult.expects, axis=-1)
> ```
> 
> maybe with additional properties to make it even simpler
> 
> ```python
> mean_states = jsseresult.mean_states
> mean_expects = jsseresult.mean_expects
> ```
The only issue as I see it is that `smart_sampling=True` and `smart_sampling=False` have the same results class. So how the expectation values and states are computed would depend on the strategy used. This information is contained in `options.smart_sampling`, so we could have a conditional in those helper methods. Would you prefer that, or to have separate results classes?

**Comment by gautierronan** [comment]
Hey @dkweiss31, I made a few minor commits on top. Apart from the two nits, I simply implemented the `mean_states` and `mean_expects` functions that we discussed previously. Indeed I'm just checking over `options.smart_sampling` to decide how to compute those: I think this should be plenty for typical usage.

I also noticed that the output when `smart_sampling=True` seems wrong on average. Using your test with `ntrajs=10_000`, I find a significant difference to the result of the ME, which is not there when `smart_sampling=False`. See the MWE here:
```python
import jax.numpy as jnp
import jax
import dynamiqs as dq
import optimistix as optx
import jax.tree_util as jtu

# parameters
ntrajs = 10000
dim = 10
smart_sampling = True # works as intended if False

# solver inputs
a = dq.destroy(dim)
H = 0.1 * a.dag() @ a + 0.4 * (a + a.dag())
jump_ops = [a, 0.3 * a.dag()]
psi0 = dq.basis(dim, 0)
tsave = jnp.linspace(0.0, 2.0, 11)
keys = jax.random.split(jax.random.key(31), num=ntrajs)
exp_ops = [a.dag() @ a]
js_options = dq.Options(smart_sampling=smart_sampling)
me_options = dq.Options(progress_meter=None)

# solve with jssesolve and mesolve
root_finder = optx.Newton(1e-4, 1e-4, jtu.Partial(optx.rms_norm))
solver = dq.method.Event(root_finder=root_finder)
jsseresult = dq.jssesolve(
    H,
    jump_ops,
    psi0,
    tsave,
    keys,
    exp_ops=exp_ops,
    options=js_options,
    method=solver,
)
meresult = dq.mesolve(H, jump_ops, psi0, tsave, exp_ops=exp_ops, options=me_options)

print(jsseresult.mean_expects.real)
print(meresult.expects.real)
```

This could be due to my `mean_expects` function (although I can't see the mistake), or to something else in the code. Maybe an issue with state normalization somewhere? Could you investage this? I quickly looked into it but didn't find anything.

@pierreguilmin Can you review the PR? Apart from this bug which should be easily fixed, it's ready for merging on my end.

**Comment by gautierronan** [comment]
https://github.com/dynamiqs/dynamiqs/pull/943/commits/934e40a361982e64b3163139077786834d83f112 might be a bit dumb actually, since it doesn't respect the variable naming anymore. But in it's previous state, it also didn't respect the overall naming conventions (capitalized single word).

**Review by gautierronan — APPROVED** [review]
Thank you for investigating and fixing @dkweiss31! On my end, the PR is good for merging. Thanks for all the work :rocket:

@pierreguilmin, here are a few points of attention when reviewing:

- https://github.com/dynamiqs/dynamiqs/commit/934e40a361982e64b3163139077786834d83f112 (variable names while printing the results class)
- https://github.com/dynamiqs/dynamiqs/pull/943/commits/2b4152c361b5e00a20f13fd81a71e616cd6427f3 (tests might be long, but previous tests failed silently)
- whether to keep `smart_sampling` to `True` or `False` by default

**Comment by pierreguilmin** [comment]
Beautiful collaborative work! I have just added two commits to this PR to align with the planned API. 😉 There are a few things I'd like to address in another PR:
- The logic of the `Event` solver is leaking (a lot!) into the API.
- We consume a random key for the deterministic no-jump trajectory, but it should be simulate independently. and should not be part of the stochastic trajectories (not included in `result.states`).
- We use the event solver to simulate the no-jump trajectory, but we should simply solve the ODE without any event handling for this part.
- We should return the no-click state probability at every time, not only the final time, otherwise the returned averages are incorrect. The shape is also incorrect, as it depends on the nb of trajectories right now.

**Comment by dkweiss31** [comment]
Thanks for the review @pierreguilmin!

> Beautiful collaborative work! I have just added two commits to this PR to align with the planned API. 😉 There are a few issues I'd like to address, I'll propose a restructuration in another PR:
> 
> * The logic of the `Event` solver is leaking (a lot!) into the API.
> * We consume a random key for the deterministic no-jump trajectory, but it should be simulate independently, and should not be part of the stochastic trajectories (not included in `result.states`).
> * We use the event solver to simulate the no-jump trajectory, but we should simply solve the ODE without any event handling for this part.

Happy with whatever modifications you'd like re these points :)
> * We should return the no-click state probability at every time, not only the final time, otherwise the returned averages are incorrect. 

I don't believe the final statement is correct (happy to return the probabilities at all times): the expectation values are computed for the normalized states as the calculation proceeds. So to calculate expectation values, the no-click trajectory should be weighted by the `p_noclick` (at the final time), and the click trajectories by `1-p_noclick`. This should yield the same results as with `improved_sampling=False`, where you obtain the no-jump trajectory with probability `p_nojump`.

>The shape is also incorrect, as it depends on the nb of trajectories right now.

Not sure I follow this? We save `final_state_norm` which indeed includes the click trajectories, but at least from its name it doesn't promise to only refer to the no-click trajectory only. Or you're referring to something different?

**Comment by pierreguilmin** [comment]
> So to calculate expectation values, the no-click trajectory should be weighted by the `p_noclick` (at the final time), and the click trajectories by `1-p_noclick`.

The probability of the no click trajectory is time dependent. The expectation value at time $t=0.1$ should not depend on the total duration of the simulation. Imagine you simulate for a total duration $T=1.0$ or $T=100.0$. The final `p_noclick` will be extremely low for a total duration $T=100.0$, and relatively high for a total duration $T=1.0$, so the expectation value estimated at time $t=0.1$ would be very different.

> Not sure I follow this? We save final_state_norm which indeed includes the click trajectories, but at least from its name it doesn't promise to only refer to the no-click trajectory only. Or you're referring to something different?

Sorry, I meant the no click probability, which should be of shape `(..., nsave)`. And I don't think we want to return the `final_state_norm` in the API (it's $1.0$ for every trajectories with a click, so useless).

**Comment by gautierronan** [comment]
I agree with @dkweiss31 on this one. In the smart sampling algorithm, sampling of the click trajectories is done according to the no-click probability, which is evaluated at the final time only. So all expectation values and states (even at short times) should be renormalized according to the final time probability.

Thus, we only care about the norm of the final no-jump state. For intermediate times, we can return it but it shouldn't be of any interest to users (and might be misleading even).
